validate.go

```diff
@@ -2,15 +2,265 @@ package main
+		deployment := appsv1.Deployment{}
+		if err := easyjson.Unmarshal(object.Request.Object, &deployment); err != nil {
+			return nil, err
+		}
+		return deployment.Metadata, nil
+	case REPLICASET_KIND:
+		replicaset := appsv1.ReplicaSet{}
+		if err := easyjson.Unmarshal(object.Request.Object, &replicaset); err != nil {
+			return nil, err
+		}
+		return replicaset.Metadata, nil
+	case STATEFULSET_KIND:
+		statefulset := appsv1.StatefulSet{}
+		if err := easyjson.Unmarshal(object.Request.Object, &statefulset); err != nil {
+			return nil, err
+		}
+		return statefulset.Metadata, nil
+	case DAEMONSET_KIND:
+		daemonset := appsv1.DaemonSet{}
+		if err := easyjson.Unmarshal(object.Request.Object, &daemonset); err != nil {
+			return nil, err
+		}
+		return daemonset.Metadata, nil
+	case REPLICATIONCONTROLLER_KIND:
+		replicationController := corev1.ReplicationController{}
+		if err := easyjson.Unmarshal(object.Request.Object, &replicationController); err != nil {
+			return nil, err
+		}
+		return replicationController.Metadata, nil
+	case CRONJOB_KIND:
+		cronjob := batchv1.CronJob{}
+		if err := easyjson.Unmarshal(object.Request.Object, &cronjob); err != nil {
+			return nil, err
+		}
+		return cronjob.Metadata, nil
+	case JOB_KIND:
+		job := batchv1.Job{}
+		if err := easyjson.Unmarshal(object.Request.Object, &job); err != nil {
+			return nil, err
+		}
+		return job.Metadata, nil
+	case POD_KIND:
+		pod := corev1.Pod{}
+		if err := easyjson.Unmarshal(object.Request.Object, &pod); err != nil {
+			return nil, err
+		}
+		return pod.Metadata, nil
+	default:
+		return nil, fmt.Errorf("object should be one of these kinds: Deployment, ReplicaSet, StatefulSet, DaemonSet, ReplicationController, Job, CronJob, Pod. Found %s", object.Request.Kind.Kind)
+	}
+}
+
+func updateLabels(currentLabels *map[string]string, currentLabelsToPropagate map[string]string) bool {
+	has_mutation := false
+	for label, newValue := range currentLabelsToPropagate {
+		if oldValue, has_label := (*currentLabels)[label]; !has_label || oldValue != newValue {
+			(*currentLabels)[label] = newValue
+			has_mutation = true
+		}
+	}
+	return has_mutation
+}
+
+func updateResourceLabels(object kubewarden_protocol.ValidationRequest, labelsToPropagate map[string]string) ([]byte, error) {
+	switch object.Request.Kind.Kind {
+	case DEPLOYMENT_KIND:
+		deployment := appsv1.Deployment{}
+		if err := easyjson.Unmarshal(object.Request.Object, &deployment); err != nil {
+			return nil, err
+		}
+		if deployment.Metadata.Labels == nil {
+			deployment.Metadata.Labels = make(map[string]string)
+		}
+		if deployment.Spec.Template.Metadata.Labels == nil {
+			deployment.Spec.Template.Metadata.Labels = make(map[string]string)
+		}
+		has_mutation := updateLabels(&deployment.Metadata.Labels, labelsToPropagate)
+		has_mutation = updateLabels(&deployment.Spec.Template.Metadata.Labels, labelsToPropagate) || has_mutation
+		if has_mutation {
+			return kubewarden.MutateRequest(deployment)
+		}
```

Comment:
It's a matter of personal taste, but I don't like this line. I would prefer something that is more straightforward and explicit:

```golang
obj_meta_mutated := updateLabels(&deployment.Metadata.Labels, labelsToPropagate)
spec_template_mutated := updateLabels(&deployment.Spec.Template.Metadata.Labels, labelsToPropagate)
if obj_meta_mutated || spec_template_mutated {
	return kubewarden.MutateRequest(deployment)
}
```
The same applies to the other snippets of code that looks like that.

Speaking of which, there's quite some repetition here:
  * check the main object metadata.labels, create a map if needed
  * check the template.metadata.labels, create a map if needed
  * check if some mutation was done, produce the final response

What about doing something like that:

```golang
switch object.Request.Kind.Kind {
case REPLICASET_KIND:
   replicaset := appsv1.ReplicaSet{}
   // unmarshal
   obj_meta_mutated := propagateLabels(&replicaset.Metadata, labelsToPropagate)
   spec_template_mutated := propagateLabels(&replicaset.Spec.Template.Metadata, labelsToPropagate)
   if obj_meta_mutated || spec_template_mutated {
      return mutate
   }
case POD_KIND:
   // similar to above, it just doesn't have the spec_template_mutated
   if mutated {
      return mutate
   }
}  
// if we got there it means no mutation happened
return accept()
```

The `propagateLabels` function would be just `updateLabels` renamed, with some extra code:

```golang
func propagateLabels(meta *Metadata, labelsToPropagate map[string]string) bool {
  if meta.Labels == nil {
     meta.Labels = make(map[string]string)
  }
  // the code of updateLabels
}
```

BTW: I just realized I'm too used to Rust, I named a bunch of variables like Rust expects :sweat_smile: